### PR TITLE
feat: Metrics utility

### PR DESCRIPTION
**Pull request by pankajagrawal16**
**Issue #, if available:**

## Description of changes:

EMF logging support

**Checklist**

 

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-java/#tenets)
* [x] Update tests
* [x] Update docs
* [x] PR title follows [conventional commit semantics]()

## Breaking change checklist

**RFC issue #**:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: powertools-metrics/src/test/java/software/amazon/lambda/powertools/metrics/handlers/PowertoolsMetricsEnabledStreamHandler.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import software.amazon.cloudwatchlogs.emf.logger.MetricsLogger;
+import software.amazon.cloudwatchlogs.emf.model.Unit;
+import software.amazon.lambda.powertools.metrics.PowertoolsMetrics;
+
+import static software.amazon.lambda.powertools.metrics.PowertoolsMetricsLogger.metricsLogger;
+
+public class PowertoolsMetricsEnabledStreamHandler implements RequestStreamHandler {
+
+    @Override
+    @PowertoolsMetrics(namespace = "ExampleApplication", service = "booking")
```
Comment (msailes) [inline powertools-metrics/src/test/java/software/amazon/lambda/powertools/metrics/handlers/PowertoolsMetricsEnabledStreamHandler.java:18]: I think this is a great developer experience!

File: example/HelloWorldFunction/src/main/java/helloworld/App.java [hunk 2]
```diff
@@ -33,12 +38,20 @@
 
     @PowertoolsLogging(logEvent = true, samplingRate = 0.7)
     @PowertoolsTracing(captureError = false, captureResponse = false)
+    @PowertoolsMetrics(namespace = "EMF", captureColdStart = true)
```
Comment (msailes) [inline example/HelloWorldFunction/src/main/java/helloworld/App.java:41]: Can you change "EMF" to something which will help customers understand what a namespace is in this context. 

"Payments" or "Fraud & Risk"

I know it's a bit of a stretch when the function is called HelloWorld. Maybe that's something we can address in the future.

File: pom.xml [hunk 3]
```diff
@@ -67,6 +68,7 @@
         <maven-source-plugin.version>3.2.1</maven-source-plugin.version>
         <maven-gpg-plugin.version>1.6</maven-gpg-plugin.version>
         <junit-jupiter.version>5.6.2</junit-jupiter.version>
+        <aws-embedded-metrics.version>0.1.0-beta</aws-embedded-metrics.version>
```
Comment (msailes) [inline pom.xml:71]: There is a v1.0.0 now 

https://repo.maven.apache.org/maven2/software/amazon/cloudwatchlogs/aws-embedded-metrics/1.0.0/

**Comment by msailes** [comment]
I'll write some Javadocs.

File: powertools-metrics/src/main/java/software/amazon/lambda/powertools/metrics/internal/LambdaMetricsAspect.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+                withSingleMetric("ColdStart", 1, Unit.COUNT, namespace(powertoolsMetrics), (logger) ->
+                        logger.setDimensions(DimensionSet.of("service", service(powertoolsMetrics)),
+                                DimensionSet.of("function_name", context.getFunctionName())));
+            }
+        }
+    }
+
+    private String namespace(PowertoolsMetrics powertoolsMetrics) {
+        return !"".equals(powertoolsMetrics.namespace()) ? powertoolsMetrics.namespace() : NAMESPACE;
```
Comment (msailes) [inline powertools-metrics/src/main/java/software/amazon/lambda/powertools/metrics/internal/LambdaMetricsAspect.java:78]: What is the default namespace if neither the annotation parameter or environment variable are set?

Comment (msailes) [reply powertools-metrics/src/main/java/software/amazon/lambda/powertools/metrics/internal/LambdaMetricsAspect.java:78]: This should probably match the "service_undefined" functionality.

File: powertools-metrics/src/main/java/software/amazon/lambda/powertools/metrics/PowertoolsMetricsLogger.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+import software.amazon.cloudwatchlogs.emf.model.Unit;
+
+public final class PowertoolsMetricsLogger {
+    private static final MetricsLogger metricsLogger = new MetricsLogger();
+
+    public static MetricsLogger metricsLogger() {
+        return metricsLogger;
+    }
+
+    public static void withSingleMetric(final String name,
```
Comment (msailes) [inline powertools-metrics/src/main/java/software/amazon/lambda/powertools/metrics/PowertoolsMetricsLogger.java:15]: Did you leave the service out on purpose?

File: powertools-metrics/src/test/java/software/amazon/lambda/powertools/metrics/handlers/PowertoolsMetricsEnabledStreamHandler.java (hunk 1 above)
Comment (pankajagrawal16) [reply powertools-metrics/src/test/java/software/amazon/lambda/powertools/metrics/handlers/PowertoolsMetricsEnabledStreamHandler.java:18]: Indeed :)

File: powertools-metrics/src/main/java/software/amazon/lambda/powertools/metrics/internal/LambdaMetricsAspect.java (hunk 4 above)
Comment (pankajagrawal16) [reply powertools-metrics/src/main/java/software/amazon/lambda/powertools/metrics/internal/LambdaMetricsAspect.java:78]: I dont see python version sets any defaults either https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/aws_lambda_powertools/metrics/base.py#L90

File: powertools-metrics/src/main/java/software/amazon/lambda/powertools/metrics/PowertoolsMetricsLogger.java (hunk 5 above)
Comment (pankajagrawal16) [reply powertools-metrics/src/main/java/software/amazon/lambda/powertools/metrics/PowertoolsMetricsLogger.java:35]: Yeah made the API consistent with python one here https://awslabs.github.io/aws-lambda-powertools-python/core/metrics/#creating-a-metric-with-a-different-dimension

**Comment by pankajagrawal16** [comment]
For some reason which I am not able to figure out yet, java doc plugin fails with error as below on java9 and 10.

```
[ERROR] /home/runner/work/aws-lambda-powertools-java/aws-lambda-powertools-java/powertools-metrics/src/main/java/software/amazon/lambda/powertools/metrics/PowertoolsMetricsLogger.java:12: warning - invalid usage of tag {@see PowertoolsMetrics}
[ERROR] javadoc: error - An exception occurred while building a component: PackageSerializedForm
[ERROR] 	(com.sun.tools.javac.code.Symbol$CompletionFailure: class file for com.fasterxml.jackson.databind.ser.impl.SimpleBeanPropertyFilter not found)
[ERROR] Please file a bug against the javadoc tool via the Java bug reporting page
[ERROR] (http://bugreport.java.com) after checking the Bug Database (http://bugs.java.com)
[ERROR] for duplicates. Include error messages and the following diagnostic in your report. Thank you.
[ERROR] com.sun.tools.javac.code.Symbol$CompletionFailure: class file for com.fasterxml.jackson.databind.ser.impl.SimpleBeanPropertyFilter not found
[ERROR] 
[ERROR] Command line was: /opt/hostedtoolcache/jdk/9.0.7/x64/bin/javadoc @options @packages
[ERROR] 
[ERROR] Refer to the generated Javadoc files in '/home/runner/work/aws-lambda-powertools-java/aws-lambda-powertools-java/powertools-metrics/target/apidocs' dir.
[ERROR] -> [Help 1]
[ERROR] 
```

**Comment by pankajagrawal16** [comment]
This commit #e0db6bc

**Comment by KassHino** [comment]
Quick question: How often does these get reviewed? And how long can these reviews take? I ask because I'm wondering if I should integrate with the EMF java library, or wait for this PR to get reviewed and merged?

**Comment by pankajagrawal16** [comment]
> Quick question: How often does these get reviewed? And how long can these reviews take? I ask because I'm wondering if I should integrate with the EMF java library, or wait for this PR to get reviewed and merged?

Hi Kassandra,

This should be released soon.

File: docs/content/core/metrics.mdx [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+This will be available in CloudWatch Logs to ease operations on high cardinal data.
+
+The `@PowertoolsMetrics` annotation **validates**, **serializes**, and **flushes** all your metrics. During metrics validation, if no metrics are provided then a warning will be logged, but no exception will be raised.
+
+If metrics are provided, and any of the following criteria are not met, `ValidationException` exception will be raised:
+
+* Minimum of 1 dimension
+* Maximum of 9 dimensions
+* Namespace is set, and no more than one
```
Comment (pankajagrawal16) [inline docs/content/core/metrics.mdx:130]: This is not something we validate explicitly

File: docs/content/core/metrics.mdx [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+* Metric units must be [supported by CloudWatch](https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_MetricDatum.html)
```
Comment (pankajagrawal16) [inline docs/content/core/metrics.mdx:131]: This is achieved by types from emf lib, there is no explicit validation.

**Review by msailes — APPROVED** [review]
Approved

**Pull request merged**